macs genome size updated and new logic to calculate it is not present

JoseEspinosa opened the pull request:
Following the discussion on #280 this PR and the response to this [issue](https://github.com/macs3-project/MACS/issues/508), this PR:
- Updates macs genome size values in the `genomes. config` as a hash containing the genome size corresponding to the following read_lengths: 50, 75,100,150 and 200. These values have been calculated as explained [here](https://deeptools.readthedocs.io/en/develop/content/feature/effectiveGenomeSize.html), as shown in the command below for a `read_length` of 50.
```console
unique-kmers.py -k 50 genome.fa
```
- Since the genome size depends on the read length a new parameter `--read_length` has been added that defaults to `null` and gives and gives an error if the pipelines is run without either providing it or the `macs_gsize` param. 
- When the `macs_gsize` value is not present in the `igenomes.config` or the user wants to overwrite it, it is still possible to provide it with `--macs_gsize` parameter.
- If any of the two options above are available, then a new local process `KHMER_UNIQUEKMERS` calculates the macs genome size and provides it to the MACS2_CALLPEAK process.


## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [ ] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [ ] `README.md` is updated

Closes #280

Emiller88 commented on nextflow_schema.json:
```diff
@@ -33,6 +33,13 @@
+                    "type": "integer",
+                    "description": "Read length to calculate macs genome size (--macs_gsize)parameter",
+                    "fa_icon": "fas fa-chart-area",
+                    "default": 100,
```
This isn't actually the default, the default is `null` currently. So it failed out of the box for me.

Emiller88 reviewed (changes requested):
Added a comment about how the docs say the default is `100` but in the `nextflow.config` the value is set to `null`.

Other than that, the code looks excellent!

Another thought, could something like [seqkit stats](https://bioinf.shenwei.me/seqkit/usage/#stats) be used to get the read_length per sample and use that? I don't know if the goal read length or the actual average is better here or not.

JoseEspinosa commented:
> Added a comment about how the docs say the default is `100` but in the `nextflow.config` the value is set to `null`.
> 
> Other than that, the code looks excellent!
> 
> Another thought, could something like [seqkit stats](https://bioinf.shenwei.me/seqkit/usage/#stats) be used to get the read_length per sample and use that? I don't know if the goal read length or the actual average is better here or not.

We also thought about it, but in the end the idea was discarded since in some cases the read length might not be uniform across samples. See discussion on slack [here](https://nfcore.slack.com/archives/CFP55NZ5G/p1656445226200299?thread_ts=1656433450.126349&cid=CFP55NZ5G). But possibly, we might come with a clever way of doing this for a subsequent release.

drpatelh commented on conf/modules.config:
```diff
@@ -503,16 +503,55 @@ if (!params.skip_plot_fingerprint) {
-            ].join(' ').trim()
+        withName: 'HOMER_ANNOTATEPEAKS_MACS2' {
+            ext.args   = '-gid'
+            // ext.prefix = 'peaks'
```
Comment needs to be dealt with here.

drpatelh replied:
Just this to sort out now?

JoseEspinosa replied:
Fix here [9f783fb](https://github.com/nf-core/chipseq/pull/283/commits/9f783fba987971a4002083f0ba0c7341df83e4ee)

drpatelh commented on lib/WorkflowChipseq.groovy:
```diff
@@ -28,6 +28,12 @@ class WorkflowChipseq {
         }
 
+        if (!params.read_length && !params.macs_gsize) {
+            log.error "Read length not specified! The pipeline requires either the '--read_length' parameter to be set" +
+            "to infer MACS2 genome size\nor to provide it with the '--macs_gsize' parameter"
```
```suggestion
            log.error "Both '--read_length' and '--macs_gsize' not specified! Please specify either to infer MACS2 genome size for peak calling."
```

drpatelh commented on lib/WorkflowChipseq.groovy:
```diff
@@ -93,8 +99,8 @@ class WorkflowChipseq {
             "  --macs_gsize parameter has not been provided.\n" +
-            "  MACS2 peak-calling and differential analysis will be skipped.\n" +
-            "  Provide '--macs_gsize macs2_genome_size' to change this behaviour.\n" +
+            "  Its value will be obtained using khmer unique-kmers.py based on the --read_length parameter.\n" +
```
```suggestion
            "  It will be auto-calculated by 'khmer unique-kmers.py' using the '--read_length' parameter.\n" +
```

drpatelh commented on nextflow.config:
```diff
@@ -40,7 +41,6 @@ params {
     // Options: Peaks
     narrow_peak                = false
     broad_cutoff               = 0.1
-    macs_gsize                 = null
```
Why was this removed? It is still a valid parameter.

drpatelh commented on nextflow_schema.json:
```diff
@@ -33,6 +33,12 @@
                 },
+                "read_length": {
+                    "type": "integer",
+                    "description": "Read length to calculate macs genome size (--macs_gsize)parameter",
```
```suggestion
                    "description": "Read length use to calculate MACS2 genome size for peak calling if `--macs_gsize` isn't provided.",
```

drpatelh reviewed (commented):
Looks great @JoseEspinosa ! Minor comments to address.

drpatelh commented on nextflow_schema.json:
```diff
@@ -33,6 +33,12 @@
                 },
+                "read_length": {
+                    "type": "integer",
+                    "description": "Read length use to calculate MACS2 genome size for peak calling if `--macs_gsize` isn't provided.",
```
```suggestion
                    "description": "Read length used to calculate MACS2 genome size for peak calling if `--macs_gsize` isn't provided.",
```

JoseEspinosa commented:
> Looks great @JoseEspinosa ! Minor comments to address.

I think I have addressed all your comments @drpatelh and @Emiller88, would be great if I can get your approval!

drpatelh reviewed (approved):
🚀
